Blinkenlights/blinkenlights.py

```diff
@@ -13,38 +13,108 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-# An exceedingly crude script to upload a bunch of png files to the matrix.
+# A script to upload a bunch of png files to the matrix.
 
+import argparse
+import contextlib
 import png
 import serial
 import sys
 
-if len(sys.argv) < 3 :
-    print("Usage: %s <gamma> <16x16 png file>..."%sys.argv[0], file=sys.stderr, );
-    exit(1)
-
-_InvGamma = 1/float(sys.argv[1])
-GammaLUT = [int(round(255*((v/255.0)**_InvGamma))) for v in range(256)]
-
-Serial = serial.Serial('/dev/ttyACM0', 9600)
-
-while True:
-    Serial.write("VER\n".encode())
-    l = Serial.readline()
-    print(l);
-    if l == b"ACK VER 1.0\r\n":
-        break
-Serial.write("ANM 600000\n".encode())
-print(Serial.readline())
-
-for png_file in sys.argv[2:]:
-    (width, height, rows, meta) = png.Reader(file=open(png_file, "rb")).asRGB8()
-    Serial.write("FRM 1000\n".encode())
-    print(Serial.readline())
-    for r in rows:
-        Serial.write(("RGB %s\n"%"".join("%02X"%GammaLUT[v] for v in r)).encode())
-        print(Serial.readline())
-Serial.write("DON\n".encode())
-print(Serial.readline())
-Serial.write("QUE\n".encode())
-print(Serial.readline())
+
+class BlinkenSerial(object):
+    def __init__(self, args):
+        self._serial = serial.Serial('/dev/ttyUSB0', 115200)
+        self._debug = args.debug
+
+        ammag = 1/args.gamma
+        self._gamma_table = [int(round(255*((v/255.0)**ammag))) for v in range(256)]
+        while True:
+            if self.command("VER") == "VER 1.0":
+                break
+
+    def write(self, line):
+        if self._debug:
+            print(f"W: {line}")
+        self._serial.write(f"{line}\n".encode())
+
+    def read(self):
+        line = self._serial.readline().decode().rstrip()
+        if self._debug:
+            print(f"R: {line:s}")
+        return line
+
+    def command(self, cmd):
+        self.write(cmd)
+        rsp = self.read().split(maxsplit=1)
+        if rsp[0] == "NAK":
+            self.write("RST")
+            self.read()
+            print(f"ERR: {rsp[1]}")
+            sys.exit(1)
+        return rsp[1]
+
+    @contextlib.contextmanager
+    def animation(self, ms):
+        self.command(f"ANM {ms}")
+        yield self
+        self.command("DON")
+
+    def frame_from_png(self, png_file, ms):
+        self.command(f"FRM {ms}")
+        rdr = png.Reader(filename=png_file)
+        (width, height, rows, meta) = rdr.asRGBA8()
+        for r in rows:
+            rgb = ""
+            for i, pixel in enumerate(r):
```

Comment:
How about using the grouper() recipe from https://docs.python.org/3/library/itertools.html#itertools-recipes to extract all four components of each pixel at a time avoiding the awkward modulo arithmetic?

```
for r, g, b, _ in grouper(r, 4, 0):
    rgb += "{0:0>2X}{1:0>2X}{2:0>2X}".format(corrected(r), corrected(g), corrected(b))
```

Reply:
I moved to using PIL (sudo apt-get install python3-pillow) instead, because it has drawing primitives. That made this problem go away too, I hope that's ok?